### ISSUE TITLE
[BUG] Two snapshotters running after network disruption

### ISSUE THREAD

**Issue by frenchtoasters** [gardener/etcd-backup-restore#517]
**Describe the bug:**
When there is a network disruption between the etcd and etcdbrctl containers it is possible to have two instances of the snapshotter taking and uploading snapshots to the snapstore. 

**Expected behavior:**
The snapshotter should always be stopped when the state moves from `Leader` to `UnknownState`. 

**How To Reproduce (as minimally and precisely as possible):**
* Create cluster
* Trigger `context deadline exceeded` when etcdbrctl connects to running etcd 
* Before the next leaderElection check interval fix connection issues
* See that snapshotter is not stopped and second one starts

**Logs:**
```
time="2022-08-03T05:20:57Z" level=info msg="Successfully saved delta snapshot at: Incr-16537142-16537653-1659504056.gz" actor=snapshotter
time="2022-08-03T05:20:57Z" level=info msg="Stopping delta snapshot..." actor=snapshotter
time="2022-08-03T05:20:57Z" level=info msg="Resetting delta snapshot to run after 5m0s." actor=snapshotter
{"level":"warn","ts":"2022-08-03T05:22:45.581Z","caller":"clientv3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"passthrough:///https://etcd-0.etcd.kube-system.svc.cluster.local:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = context deadline exceeded"}
time="2022-08-03T05:22:45Z" level=error msg="failed to get status of etcd endPoint: https://etcd-0.etcd.kube-system.svc.cluster.local:2379 with error: context deadline exceeded" actor=leader-elector
time="2022-08-03T05:22:45Z" level=error msg="failed to elect the backup-restore leader: context deadline exceeded" actor=leader-elector
time="2022-08-03T05:22:45Z" level=info msg="backup-restore stops leading..." actor=backup-restore-server
time="2022-08-03T05:22:45Z" level=info msg="Setting status to : 503" actor=backup-restore-server
time="2022-08-03T05:22:45Z" level=info msg="backup-restore is in: UnknownState" actor=leader-elector
time="2022-08-03T05:22:45Z" level=info msg="waiting for Re-election..." actor=leader-elector
time="2022-08-03T05:22:45Z" level=info msg="Closing defragmentor." actor=backup-restore-server
time="2022-08-03T05:22:45Z" level=info msg="Snapshotter stopped." actor=backup-restore-server
time="2022-08-03T05:22:45Z" level=info msg="Probing etcd..." actor=backup-restore-server
time="2022-08-03T05:22:45Z" level=info msg="Shutting down..." actor=backup-restore-server
time="2022-08-03T05:22:45Z" level=info msg="GC: Stop signal received. Closing garbage collector." actor=snapshotter
time="2022-08-03T05:22:51Z" level=info msg="backup-restore became: Leader" actor=leader-elector
time="2022-08-03T05:22:51Z" level=info msg="backup-restore started leading..." actor=leader-elector
time="2022-08-03T05:22:51Z" level=info msg="Creating snapstore from provider: S3" actor=backup-restore-server
time="2022-08-03T05:22:51Z" level=info msg="Creating snapshotter..." actor=backup-restore-server
time="2022-08-03T05:22:55Z" level=info msg="Probing etcd..." actor=backup-restore-server
time="2022-08-03T05:22:55Z" level=info msg="Setting status to : 200" actor=backup-restore-server
time="2022-08-03T05:22:55Z" level=info msg="Applied watch on etcd from revision: 16537654" actor=snapshotter
{"level":"warn","ts":"2022-08-03T05:23:05.572Z","caller":"clientv3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"passthrough:///https://etcd-0.etcd.kube-system.svc.cluster.local:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = context deadline exceeded"}
time="2022-08-03T05:23:05Z" level=error msg="failed to get status of etcd endPoint: https://etcd-0.etcd.kube-system.svc.cluster.local:2379 with error: context deadline exceeded" actor=leader-elector
time="2022-08-03T05:23:05Z" level=error msg="failed to elect the backup-restore leader: context deadline exceeded" actor=leader-elector
time="2022-08-03T05:23:05Z" level=info msg="Setting status to : 503" actor=backup-restore-server
time="2022-08-03T05:23:05Z" level=info msg="backup-restore is in: UnknownState" actor=leader-elector
time="2022-08-03T05:23:05Z" level=info msg="waiting for Re-election..." actor=leader-elector
time="2022-08-03T05:23:05Z" level=info msg="Closing defragmentor." actor=backup-restore-server
time="2022-08-03T05:23:10Z" level=info msg="backup-restore became: Leader" actor=leader-elector
time="2022-08-03T05:23:10Z" level=info msg="backup-restore started leading..." actor=leader-elector
time="2022-08-03T05:23:10Z" level=info msg="Creating snapstore from provider: S3" actor=backup-restore-server
time="2022-08-03T05:23:10Z" level=info msg="Creating snapshotter..." actor=backup-restore-server
time="2022-08-03T05:23:11Z" level=info msg="Probing etcd..." actor=backup-restore-server
time="2022-08-03T05:23:11Z" level=info msg="Taking delta snapshot for time: 2022-08-03 05:23:11.192794202 +0000 UTC" actor=snapshotter
time="2022-08-03T05:23:11Z" level=info msg="start the Compression of delta snapshot" actor=snapshotter
time="2022-08-03T05:23:11Z" level=info msg="start compressing the snapshot using gzip Compression Policy" actor=compressor
time="2022-08-03T05:23:11Z" level=info msg="Total written bytes: 1016074" actor=compressor
time="2022-08-03T05:23:11Z" level=info msg="Setting status to : 200" actor=backup-restore-server
time="2022-08-03T05:23:11Z" level=info msg="Successfully initiated the multipart upload with upload ID : 2~2WI_MI8F4MzplcdHn-3_BkXAP__UiuH"
time="2022-08-03T05:23:11Z" level=info msg="Uploading snapshot of size: 271291, chunkSize: 5242880, noOfChunks: 1"
time="2022-08-03T05:23:11Z" level=info msg="Triggered chunk upload for all chunks, total: 1"
time="2022-08-03T05:23:11Z" level=info msg="No of Chunks:= 1"
time="2022-08-03T05:23:11Z" level=info msg="Uploading chunk with id: 1, offset: 0, attempt: 0"
time="2022-08-03T05:23:11Z" level=info msg="Applied watch on etcd from revision: 16537654" actor=snapshotter
time="2022-08-03T05:23:11Z" level=info msg="Taking delta snapshot for time: 2022-08-03 05:23:11.993767293 +0000 UTC" actor=snapshotter
time="2022-08-03T05:23:11Z" level=info msg="start the Compression of delta snapshot" actor=snapshotter
time="2022-08-03T05:23:11Z" level=info msg="start compressing the snapshot using gzip Compression Policy" actor=compressor
time="2022-08-03T05:23:12Z" level=info msg="Total written bytes: 1104441" actor=compressor
time="2022-08-03T05:23:12Z" level=info msg="Received chunk result for id: 1, offset: 0"
time="2022-08-03T05:23:12Z" level=info msg="Received successful chunk result for all chunks. Stopping workers."
time="2022-08-03T05:23:12Z" level=info msg="Finishing the multipart upload with upload ID : 2~2WI_MI8F4MzplcdHn-3_BkXAP__UiuH"
time="2022-08-03T05:23:12Z" level=info msg="Successfully initiated the multipart upload with upload ID : 2~0rJjtqi_y08xm4sf6QMycSYi5OAHW_P"
time="2022-08-03T05:23:12Z" level=info msg="Uploading snapshot of size: 291902, chunkSize: 5242880, noOfChunks: 1"
time="2022-08-03T05:23:12Z" level=info msg="Triggered chunk upload for all chunks, total: 1"
time="2022-08-03T05:23:12Z" level=info msg="No of Chunks:= 1"
time="2022-08-03T05:23:12Z" level=info msg="Uploading chunk with id: 1, offset: 0, attempt: 0"
time="2022-08-03T05:23:12Z" level=info msg="Total time to save delta snapshot: 1.469544 seconds."
time="2022-08-03T05:23:12Z" level=info msg="Successfully saved delta snapshot at: Incr-16537654-16537842-1659504191.gz" actor=snapshotter
time="2022-08-03T05:23:12Z" level=info msg="Starting the garbage collector..." actor=backup-restore-server
time="2022-08-03T05:23:12Z" level=info msg="Starting snapshotter..." actor=backup-restore-server
time="2022-08-03T05:23:12Z" level=info msg="Will take next full snapshot at time: 2022-08-03 06:00:00 +0000 UTC" actor=snapshotter
time="2022-08-03T05:23:13Z" level=info msg="Received chunk result for id: 1, offset: 0"
time="2022-08-03T05:23:13Z" level=info msg="Received successful chunk result for all chunks. Stopping workers."
time="2022-08-03T05:23:13Z" level=info msg="Finishing the multipart upload with upload ID : 2~0rJjtqi_y08xm4sf6QMycSYi5OAHW_P"
time="2022-08-03T05:23:13Z" level=info msg="Total time to save delta snapshot: 1.602461 seconds."
time="2022-08-03T05:23:13Z" level=info msg="Successfully saved delta snapshot at: Incr-16537654-16537858-1659504191.gz" actor=snapshotter
```

**Screenshots (if applicable):**

**Environment (please complete the following information):**
 - Etcd version/commit ID : bitnami/etcd:3.4.14
 - Etcd-backup-restore version/commit ID: v0.15.0+
 - Cloud Provider [All/AWS/GCS/ABS/Swift/OSS]: ALL

**Anything else we need to know?**:

**Comment by ishan16696** [comment]
Hi @frenchtoasters .
I have doubts regarding the steps you mentioned to reproduce this issue:

> Trigger context deadline exceeded when etcdbrctl connects to running etcd

when we trigger the context deadline exceeded then backup-restore will move to leader -> unknown state, and will try to perform [some action](https://github.com/gardener/etcd-backup-restore/blob/master/pkg/leaderelection/leaderelection.go#L101) like closing snapshotter. Is this [check interval](https://github.com/gardener/etcd-backup-restore/blob/master/pkg/leaderelection/leaderelection.go#L87) should large enough, so we can be sure that all action have been completed ?

**Comment by frenchtoasters** [comment]
Hey @ishan16696 
> Is this [check interval](https://github.com/gardener/etcd-backup-restore/blob/master/pkg/leaderelection/leaderelection.go#L87) should large enough, so we can be sure that all action have been completed?

We are taking the defaults for those values in our deployment. 

The only thing that we can tell, as the log suggests, that when there is _two_ `context deadline exceeded` messages that pop up within quick succession this issue will occur. 

Looking at the timestamps between the errors below, this time it happened within 2 minutes. 
```
time="2022-08-03T05:20:57Z" level=info msg="Resetting delta snapshot to run after 5m0s." actor=snapshotter
{"level":"warn","ts":"2022-08-03T05:22:45.581Z","caller":"clientv3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"passthrough:///https://etcd-0.etcd.kube-system.svc.cluster.local:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = context deadline exceeded"}
```
```
time="2022-08-03T05:22:55Z" level=info msg="Applied watch on etcd from revision: 16537654" actor=snapshotter
{"level":"warn","ts":"2022-08-03T05:23:05.572Z","caller":"clientv3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"passthrough:///https://etcd-0.etcd.kube-system.svc.cluster.local:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = context deadline exceeded"}
time="2022-08-03T05:23:05Z" level=error msg="failed to get status of etcd endPoint: https://etcd-0.etcd.kube-system.svc.cluster.local:2379 with error: context deadline exceeded" actor=leader-elector
```

After the second one we can clearly see that the snapshotter didnt close within 5 seconds of the defragmentor, and in that time the pod became the leader again and thus started a new snapshotter. 
```
time="2022-08-03T05:23:05Z" level=info msg="backup-restore is in: UnknownState" actor=leader-elector
time="2022-08-03T05:23:05Z" level=info msg="waiting for Re-election..." actor=leader-elector
time="2022-08-03T05:23:05Z" level=info msg="Closing defragmentor." actor=backup-restore-server
time="2022-08-03T05:23:10Z" level=info msg="backup-restore became: Leader" actor=leader-elector
time="2022-08-03T05:23:10Z" level=info msg="backup-restore started leading..." actor=leader-elector
time="2022-08-03T05:23:10Z" level=info msg="Creating snapstore from provider: S3" actor=backup-restore-server
time="2022-08-03T05:23:10Z" level=info msg="Creating snapshotter..." actor=backup-restore-server
```

> when we trigger the context deadline exceeded then backup-restore will move to leader -> unknown state, and will try to perform [some action](https://github.com/gardener/etcd-backup-restore/blob/master/pkg/leaderelection/leaderelection.go#L101) like closing snapshotter.

I agree that appears to be the path that things are going down however based on the logs that we have something in those call backs is hanging before the message is sent to stop the snapshotter. Which to me seems like some type of race condition with likely the mutex as that is the only thing that is acted on before sending the stop signal.

**Comment by ishan16696** [comment]
Although I'm unable to reproduce this issue but I think this PR:https://github.com/gardener/etcd-backup-restore/pull/537 might also solve this issue as well.

**Comment by ishan16696** [comment]
Apart fromt this PR:https://github.com/gardener/etcd-backup-restore/pull/537 se have introduced some more improvements in closing of snapshottter with this PR:https://github.com/gardener/etcd-backup-restore/pull/546 in backup-restore v0.21.0.
Please let me know if you observe this issue again.

**Comment by ishan16696** [comment]
With these PRs: https://github.com/gardener/etcd-backup-restore/pull/537 and https://github.com/gardener/etcd-backup-restore/pull/546 which fixed race-conditions in backup-restore `v0.21.0` , I haven’t observed this behaviour of running 2 snapshotters.
For now, I'm closing this issue. Feel free to reopen it if you see it again.
/close
